### PR TITLE
Server context path for Spring Flux

### PR DESCRIPTION
**Pull request by gtukmachev**
support of the `server.servlet.context-path` property for WebFlux

tested with standard spring-web applications
tested with WebFlux with the following setup of root context path:

```
application.yaml
server:
    port: 8080
    servlet.context-path: '/api'
```
```
// Spring Boot Application 

@SpringBootApplication
public class Application {

    public static void main(String[] args) {
        run(Application.class, args);
    }

    @Bean
    public UndertowReactiveWebServerFactory undertowReactiveWebServerFactory(
            @Value("${server.servlet.context-path}") String contextPath
    ) {
        return new UndertowReactiveWebServerFactory() {
            @Override
            public WebServer getWebServer(HttpHandler httpHandler) {
                Map<String, HttpHandler> handlerMap = new HashMap<>();
                handlerMap.put(contextPath, httpHandler);
                return super.getWebServer(new ContextPathCompositeHandler(handlerMap));
            }
        };
    }

}
```

### CODE REVIEW

**Comment by gtukmachev** [comment]
hmm.. the test `de.codecentric.boot.admin.server.notify.RemindingNotifierTest#should_resubscribe_after_error` is failed on Travis, but, I think, I didn't make changes to touch any functionality for this test.

Locally the test passed successfully as well (the only difference I can see - Travis uses a Unix based OS, but I have Win 10)

**Comment by joshiste** [comment]
If I get your config right, you are using the "standard"-property `server.servlet.context-path` to set the path for your reactive handler in your custom code. So this is not standard spring boot behavior or is it?

**Comment by joshiste** [comment]
Imho a solution proposed in #804 would be a much better fit for this problem.

**Comment by gtukmachev** [comment]
Hi, Johannes. Thanks a lot for your activity!

> If I get your config right, you are using the "standard"-property server.servlet.context-path to set the path for your reactive handler in your custom code. So this is not standard spring boot behavior or is it?

You are right!  Unfortunately, I didn't find a proper standard way to configure a 'root contect path'  for reactive API in case of using MVC Controllers. The best I found (see my example with `UndertowReactiveWebServerFactory`  bean) helped me to solve my problem.... and I choose to use the standard property for the configuration.

> Imho the solution proposed in #804 would be a much better fit for this problem.

I can update my solution... actually - adding a new property under `spring-admin` namespace was my first idea, but I've chosen such way because it described In `spring-boot-client` documentation: http://codecentric.github.io/spring-boot-admin/2.0.3/#spring-boot-admin-client (see `spring.boot.admin.client.instance.service-url` property description).
```
Guessed based on service-base-url and server.context-path.
```

So, I made my solution to meet the documentation. 

Let's discuss the best way and I`ll implement it. 
I suggest: 
1. add the new property (from #804) - `spring.boot.admin.client.instance.service-path`
in y solution change using of the  `server.servlet.context-path` to this new one (so - both reactive and servlet branches will have the same implementation)
2. as  a default value for this new property - use the standard `server.servlet.context-path` - we'll save "as-is" the behavior for the non-reactive branch, and we'll provide the ability to configure it for the reactive branch as well.

What do you think?

**Comment by joshiste** [comment]
>  I can update my solution... actually - adding a new property under spring-admin namespace was my first idea, but I've chosen such way because it described In spring-boot-client documentation: http://codecentric.github.io/spring-boot-admin/2.0.3/#spring-boot-admin-client (see spring.boot.admin.client.instance.service-url property description).

The docs are misleading and outdated as this property doesn't exist in spring boot 2.

File: spring-boot-admin-client/src/main/java/de/codecentric/boot/admin/client/registration/DefaultApplicationFactory.java [hunk 1]
```diff
@@ -84,7 +84,22 @@ protected String getServiceUrl() {
...
+    protected String getServicePath() {
+        if (instance.getServicePath() != null) {
+            return instance.getServicePath();
+        }
+
+        return getServerContextPath();
+    }
+
+    protected String getServerContextPath() {
+        return server.getServlet().getContextPath();
```
Comment (joshiste) [inline spring-boot-admin-client/src/main/java/de/codecentric/boot/admin/client/registration/DefaultApplicationFactory.java]: Imho `server.getServlet().getContextPath();` should be moved to the `ServletApplicationFactory`and replaced with `/`.

File: spring-boot-admin-client/src/main/java/de/codecentric/boot/admin/client/registration/DefaultApplicationFactory.java [hunk 2]
```diff
@@ -84,7 +84,22 @@ protected String getServiceUrl() {
             return instance.getServiceUrl();
         }
 
-        return UriComponentsBuilder.fromUriString(getServiceBaseUrl()).path("/").toUriString();
+        return UriComponentsBuilder.fromUriString(getServiceBaseUrl())
+                                   .path("/")
```
Comment (joshiste) [inline spring-boot-admin-client/src/main/java/de/codecentric/boot/admin/client/registration/DefaultApplicationFactory.java]: Shouldn't this line be removed?

**Comment by joshiste** [comment]
@gtukmachev will you update the PR?

**Comment by joshiste** [comment]
closed due to missing feedback

**Comment by leshalv** [comment]
image

I think this solves the Server context path problem. Can I submit a PR? But this trial only works with Spring Boot 2x columns.

**Pull request closed without merging**